app/models/rental.rb

```diff
@@ -0,0 +1,47 @@
+class Rental < ApplicationRecord
+  belongs_to :customer
+  belongs_to :movie
+
+  validates :due_date, presence: true
+  validate :valid_due_date
+  validate :check_inventory
+
+  def initialize rental_params
+    converted_date = convert_date(rental_params[:due_date])
+    rental_params[:due_date] = converted_date
+    movie =  Movie.find_by(title: rental_params[:title])
+    if movie
+      rental_params[:movie_id] = movie.id
+    end
+    rental_params.delete(:title)
+    super(rental_params)
+  end
+
+  def check_inventory
+    if movie
+      if movie.available_inventory <= 0
+        rentals = Rental.where(movie_id: movie.id)
+        earliest = rentals.min_by { |rental| rental.due_date}
+        errors.add(:availability, "Sorry, this movie is not in stock. One is due back on #{earliest.due_date}")
+      end
+    end
+  end
+
+  def convert_date date
+    if date =~ /\A\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])\z/
+      return Date.parse(date)
+    else
+      date
+    end
+  end
+
+  def valid_due_date
+    if due_date.class != Date
```

Comment:
How do you take in the due date? Would checking if the due date is too far in the future (like if I tried to check something out for 2183721983 years) be necessary?